tests/correctness/control_flow.rs

```diff
@@ -483,6 +483,86 @@ fn for_loop_and_increment_10_times_skipping_1() {
     assert_eq!(res, 1005);
 }
 
+#[test]
+fn for_loop_does_not_execute_if_condition_is_not_met() {
+    let function = r#"
+    FUNCTION main : DINT
+    VAR
+        i, end, step, res: DINT;
+    END_VAR
+        end := -1;
+        step := -1;
+        res := 100;
+
+        FOR i := end TO 0 BY step DO
+            res := i;
+        END_FOR
+
+        main := res;
+    END_FUNCTION
+    "#;
+
+    let res: i32 = compile_and_run(function.to_string(), &mut crate::MainType::default());
+    assert_eq!(res, 100);
+}
+
+#[test]
+fn for_loop_step_changes_sign_in_loop_body() {
+    let function = r#"
+    FUNCTION main : DINT
+    VAR
+        i, step, temp, iteration: DINT;
+    END_VAR
+        step := 1;
+        iteration := 0;
+
+        FOR i := 5 TO 10 BY step DO
+            temp := (step + 1) * -1 ;
+            IF i + temp > 0 THEN
+                step := temp;
+            ELSE
+                step := step + 3;
+            END_IF;
+
+            // i:     5, 3,  4, 2,  3, 1, 2,  6, 1, 5, 12
+            // step: -2, 1, -2, 1, -2, 1, 4, -5, 4, 7
+            iteration := iteration + 1;
+        END_FOR
+        
+        main := iteration;
+    END_FUNCTION
+    "#;
+
+    let res: i32 = compile_and_run(function.to_string(), &mut crate::MainType::default());
+    assert_eq!(res, 10);
+}
+
+#[test]
+fn for_loop_step_and_counter_change_sign_in_loop_body() {
+    let function = r#"
+    FUNCTION main : DINT
+    VAR
+        i, step, temp, iteration: DINT;
+    END_VAR
+        step := 1;
+        iteration := 0;
+
+        FOR i := 3 TO 10 BY step DO
+            step := (step + 1) * -2;
+
+            // i:     3, -1,   5, -9, 17
+            // step: -4,  5, -14, 26
+            iteration := iteration + 1;
+        END_FOR
+        
+        main := iteration;
+    END_FUNCTION
```

Comment:
I've added two tests with edge cases, were the `STEP` variable changes signage in the loop body. Do we expect the loop to continue until `COUNTER >= END` if the loop is determined as incrementing (current behaviour)? 
Should we also exit the loop if `COUNTER <= START` for incrementing loops (in which case we'd only expect 1 iteration in the test above)?

Reply:
Personally I would just mimic what the C language would do in these edge-cases; I suspect it will just check for the condition `COUNTER >= END` and only stop if it becomes true

Reply:
out of interest, what is codesys doing in these cases? is the step by constant? It might be that we always have to invert the check based on the sign.
C has the exit condition defined in the for loop, so it's not the same case.

Reply:
yeah, mimicking C is not an ideal solution here unfortunately.  I have already tried changing the predicate dynamically based on the sign. This results in most loops executing once, e.g.:
```
FOR i := 0 TO 10 BY x DO
    invert x and check
```
If x starts out as let's say `2`, upon second iteration the check will be `IF 2 >= 10`. The exception here is if `step` is larger than the absolute difference between `start` and `end`. 
I think checking codesys behaviour is the way to go here, too.
Is this a blocker for this story? I don't think this was accounted for previously and might be out of scope for this bug fix. I'd suggest tracking it separately. @ghaith @volsa thoughts?